### PR TITLE
Standardizing tests

### LINKED ISSUE #448: Move all shared golden functionalities to testutils
We should move all shared golden functionalities to testutils:
- functions that updates golden
- remove the tree manual comparison in some tests
- install the update flag from testutils
- Reshape CompareTreesForFiltering so update with the update flag. However, it’s used a couple of times to only compare, so care should be taken.

### LINKED ISSUE #452: Standardize in capital/non capital letters for subtests
We are not consistent for now here.

### LINKED ISSUE #453: Golden should always be normalized
Golden file generation should use the same helper for normalizing the golden name.

We need then to discuss if we shouldn’t have them by test group name, like `TestNew/Subtest/golden` or `golden/TestNew/Subtest`. This will need to take into account we have some fixtures per tests sometimes.

### PR DESCRIPTION
**Pull request by denisonbarbosa**
Oof, finally...

This PR aims to do some house cleaning and standardize some things in our tests:

1) The main objective: testutils and golden files. Most of the golden files manipulation code was moved to testutils, reducing the need to this kind of business inside test code. This includes some package functions that had potential to be used elsewhere. Also, each golden file had their name normalized to maintain consistency.

2) Standardizing test case nomenclature. As voted, our test cases now follow the  "This is a test case" syntax, aside from a few that were involved in too many shenanigans on testdata.

Yes, I did a commit for each package / subpackage tests to make it easier to see the diff and fix some issues if needed.

Closes #453 
Closes #452 
Closes #448 (Maybe?)

### CODE REVIEW

**Review by didrocks — CHANGES_REQUESTED** [review]
Very good work as said already on MM :) I have few comments about missed spot and how we should use some part of the API, but this is a really great improvements over all and we should then force to follow that structure for everything.

It’s also great to see the amount of code duplication which is now out of the window :)
I did a last global tree check and I find it way easier to see common fixtures, fixtures specific to a test family and golden files well ordered. (oh, instead of main test, what about "test family" to name the super test, I’m afraid if I tried to edit the comment after finding it, it will vanish too :p ?).

So, again, in a nutshell, very nice work!
Please, for review easyness, do you mind appending new comits for changes and only rebase once you get the final approval? That way, I don’t have to look again at the whole diff and make my webrowser suffering too much :p

File: internal/watchdtui/watchdtui_test.go
Comment (denisonbarbosa) [reply internal/watchdtui/watchdtui_test.go:387]: No, because we changed directory earlier in the function and this would be bad. Believe me, I made this mistake earlier 🥲

File: internal/policies/scripts/scripts_test.go
Comment (didrocks) [reply internal/policies/scripts/scripts_test.go:114]: oh sorry, on the second reading, I missed wantErr: false on that one. I agree on special casing and separate it (so maybe above errors stenza?) as it clearly (as you can see), introduced some confusion here when skimming over it :)

File: internal/watchdtui/watchdtui_test.go
Comment (didrocks) [reply internal/watchdtui/watchdtui_test.go:387]: Mind adding a comment on the assignement then telling we are capturing the current path here because we are changing the directory afterwards? Because I can really forsee some quick cleanup which will do exactly that :p
(maybe we should then rename as it’s not absPath anymore, it’s absPackageDirPath or something like this)

File: internal/testutils/golden.go
Comment (denisonbarbosa) [reply internal/testutils/golden.go:61]: This is a weird state, I agree. The thing is: as we discussed on MM, the idea was to keep the LoadFromGolden dealing in strings only, but later down the line (in the admxgen package) there was a whole parallel set of golden functionalities implemented inside the tests and I decided to move them to testutils, where they belong.
Also, despite having the logic kind of implemented twice, I kind of like having it for two reasons:
- Serialization with strings do work, but will break the ones that have color tags, so that's a no no. However, we really benefit from having a generic one that is able to serialize structs, slices, etc. (specially in the long term).
- Having both allows us to use the string one for strings, which is faster and used by a lot of our tests, and keep the generics one for more complex datatypes that would require serialization to be written down.

But I totally agree the name is weird and that I should figure out a way of stopping it from being called to strings.

Comment (didrocks) [reply internal/testutils/golden.go:61]: I’ll give the refactoring a go after this PR I think to reuse the string one. I’m happy to give it a try. Agreed that we should change the name anyway in that PR.

Comment (denisonbarbosa) [reply internal/testutils/golden.go:61]: I could use reflect to try and check if the provided E needs serialization, wdyt?

Comment (didrocks) [reply internal/testutils/golden.go:61]: I was thinking about the other way around, I’ll try this afternoon to show you what I meant

File: internal/testutils/golden.go
Comment (didrocks) [inline internal/testutils/golden.go:94]: can you avoid defining path and just return it?

File: internal/testutils/golden.go
Comment (didrocks) [reply internal/testutils/golden.go]: > I could use reflect to try and check if the provided E needs serialization, wdyt?

Actually I would restrict to what the yaml marhsall/unmarshall use. As it’s interface{} for now, let’s not bother in the end. The other changes looks good!
I’m sad we can’t override the same function (one for string, the other for "the rest") and keep a single function name, but this is not doable (which is understandable: what happens if you have 2 functions using generic where the constraints could work with this type?

File: internal/authorizer/authorizer_test.go
Comment (didrocks) [inline internal/authorizer/authorizer_test.go:49]: I would add to the end something like "is unauthorized" to both tests while we are enhancing them.

**Comment by didrocks** [comment]
2 smalls comments remaining and then, we are done! :)

**Review by didrocks — APPROVED** [review]
All good now! Thanks for doing the changes and takcling this one.

Feel free to merge once you reorganized the commits.

**Review by didrocks — APPROVED** [review]
I did a diff with previous pre-rebase and I only see you rebased on main, so looks good!

**Pull request merged**
